### PR TITLE
Upgrade the cpf calculator to use the CPF ceiling for additional wages.

### PR DESCRIPTION
**Pull request by bramski**
**Feature:**
This will add the ```ytd_ow_subject_to_cpf``` and ```ytd_additional_wages``` to the calculator where the year's cumulative earnings may be passed.

It also now returns ```ow_subject_to_cpf``` and ```aw_subject_to_cpf``` which are the ordinary and additional wages which were subject to cpf.  These must be tracked and passed into the calculator for continuing yearly payroll.

**Ticket:**
https://payrollhero5161.leankit.com/Boards/View/253241089/279860052/310977481

**Submitter Checklist:**
- [ ] CI Passing
- [ ] No Code Climate Issues
- [ ] Class, Module, and public method comments up to spec

**Reviewer Checklist:**
- [ ] New Code Tested Effectively
- [ ] Code organization meets SOP standards
- [ ] Does this code fulfill the ticket?

**Only Then:**
- [ ] Move card in Leankit

### CODE REVIEW

File: lib/singapore_cpf_calculator.rb [hunk 1]
```diff
@@ -34,23 +35,24 @@ class << self
...
-      validate_params date: date,
+    def calculate(date:,
+                  birthdate:,
+                  residency_status:,
+                  spr_start_date: nil,
+                  ordinary_wages:,
+                  additional_wages:,
+                  employee_contribution_type: nil,
+                  employer_contribution_type: nil,
+                  cumulative_ordinary: 0.0)
```
Comment (more-ron) [inline lib/singapore_cpf_calculator.rb]: Not sure what's best to call this `cumulative_ordinary`. Maybe `year_total_ordinary_wages`. 

Also, I thought we are doing the `year_total_additional_wages` ceiling and not the ordinary ceiling?

Comment (more-ron) [inline lib/singapore_cpf_calculator.rb]: The name of the ticket is: {4} [M - M] Implement the Additional Wage ceiling for CPF https://payrollhero5161.leankit.com/Boards/View/253241089/279878982

File: lib/singapore_cpf_calculator/base_calculator.rb [hunk 2]
```diff
@@ -61,6 +64,16 @@ def total_wages
       ordinary_wages + additional_wages
     end
 
+    def clip_additional_wages_based_on_cumulative
+      if additional_wage_ceiling.blank?
+        additional_wages
+      elsif cumulative_ordinary >= additional_wage_ceiling
+        d('0.0')
+      else
+        [additional_wages, additional_wage_ceiling - cumulative_ordinary].min
```
Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Maybe extract `additional_wage_ceiling - cumulative_ordinary` to an aptly named variable/method so it's easier to get a context. 

```ruby
[additional_wages, within_ceiling_additional_wages]
```

Also, `additional_wage_ceiling - cumulative_ordinary` could be negative and will always be the minimum after the ceiling is achieved.

The within ceiling value is actually zero or any positive value.

```ruby
def within_ceiling_additional_wages
  [BigDecimal(0), additional_wage_ceiling - year_total_additional_wages].max
end
```

File: lib/singapore_cpf_calculator.rb (hunk 1 above)
Comment (bramski) [inline lib/singapore_cpf_calculator.rb]: Trying to keep the name short.  The parameter has a description.  I don't want to find/replace the whole project.

I thought the ceiling was based upon ordinary wages?

Comment (bramski) [inline lib/singapore_cpf_calculator.rb]: screen shot 2016-03-14 at 5 14 12 pm

File: lib/singapore_cpf_calculator.rb [hunk 3]
```diff
@@ -12,7 +12,8 @@ class << self
     #                                             ordinary_wages: 700.00,
     #                                             additional_wages: 252.00,
     #                                             employee_contribution_type: "full",
-    #                                             employer_contribution_type: "full"
+    #                                             employer_contribution_type: "full",
+    #                                             cumulative_ordinary: 0.0,
```
Comment (more-ron) [inline lib/singapore_cpf_calculator.rb]: Update documentation. There should also be a readme file that needs the same.

File: lib/singapore_cpf_calculator/base_calculator.rb [hunk 4]
```diff
@@ -30,20 +33,24 @@ def required_age_group
...
+      @additional_wages = clip_additional_wages_based_on_ceiling(additional_wages)
     end
 
     # @return [Hash] returns the total, employee, employer amounts for the CPF contribution
     def calculate
       CPFContribution.new total: total_contribution,
-                          employee: employee_contribution
+                          employee: employee_contribution,
+                          aw_subject_to_cpf: additional_wages,
+                          ow_subject_to_cpf: capped_ordinary_wages
```
Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Where's the `capped_ordinary_wages` coming from?

File: lib/singapore_cpf_calculator/base_calculator.rb [hunk 5]
```diff
@@ -61,6 +68,19 @@ def total_wages
...
+    def clip_additional_wages_based_on_ceiling(additional_wages)
+      if additional_wage_ceiling.blank?
+        additional_wages
+      else
+        [additional_wages, calculated_remaining_wage_ceiling].min
+      end
+    end
+
+    def calculated_remaining_wage_ceiling
+      max_remaining = additional_wage_ceiling - ytd_ow_subject_to_cpf - capped_ordinary_wages - ytd_additional_wages
```
Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Maybe call this `remaining_additional_wages_subject_to_cpf`?

Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: `current_maximum_additional_wages_subject_to_cpf ` ?

File: lib/singapore_cpf_calculator/base_calculator.rb (hunk 4 above)
Comment (bramski) [inline lib/singapore_cpf_calculator/base_calculator.rb]: That's the ordinary wages that were subject to CPF.  Not sure what you mean by "coming from".  It's a member method.

File: lib/singapore_cpf_calculator/base_calculator.rb (hunk 5 above)
Comment (bramski) [inline lib/singapore_cpf_calculator/base_calculator.rb]: It is exactly what it says it is.  I'm not changing the name.  It's the unused wage ceiling.

Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: I'm a bit confused about the formula:

```ruby
current_value = ceiling - projected_current_year_ordinary_wages - ytd_additional_wages
```

I guess?

```ruby
projected_current_year_ordinary_wages = ytd_ow_subject_to_cpf + capped_ordinary_wages
```

Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Might be best to rename the variables so they're easy to read.

File: lib/singapore_cpf_calculator/base_calculator.rb (hunk 4 above)
Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Ah, got ya.

File: lib/singapore_cpf_calculator/base_calculator.rb (hunk 5 above)
Comment (more-ron) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Got it. This is the other option that's not doing the projected ordinary wages value.

**Comment by more-ron** [comment]
Looks good to :shipit:

File: lib/singapore_cpf_calculator/base_calculator.rb (hunk 5 above)
Comment (bramski) [inline lib/singapore_cpf_calculator/base_calculator.rb]: Yeah there's no projection going on here.  We just keep a running tally

**Pull request merged**